io/zenoh-transport/src/unicast/universal/tx.rs

```diff
@@ -11,51 +11,106 @@
 // Contributors:
 //   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
 //
-use zenoh_core::zread;
-use zenoh_protocol::network::NetworkMessage;
+use zenoh_protocol::{
+    core::{Priority, PriorityRange, Reliability},
+    network::NetworkMessage,
+};
 
 use super::transport::TransportUnicastUniversal;
 #[cfg(feature = "shared-memory")]
 use crate::shm::map_zmsg_to_partner;
+use crate::unicast::transport_unicast_inner::TransportUnicastTrait;
 
 impl TransportUnicastUniversal {
-    fn schedule_on_link(&self, msg: NetworkMessage) -> bool {
-        macro_rules! zpush {
-            ($guard:expr, $pipeline:expr, $msg:expr) => {
-                // Drop the guard before the push_zenoh_message since
-                // the link could be congested and this operation could
-                // block for fairly long time
-                let pl = $pipeline.clone();
-                drop($guard);
-                tracing::trace!("Scheduled: {:?}", $msg);
-                return pl.push_network_message($msg);
-            };
+    /// Returns the index of the best matching [`Reliability`]-[`PriorityRange`] pair.
```

Comment:
The 'smaller range' rule prevents us from returning as soon as we find a matching range. Is it worth it ?
Even with this rule we could return as soon as we find a full match with a range of 1 (cannot find a smaller range later).

Reply:
Sorry to disappoint but the search isn't even short-circuiting. It searches for three kinds of matches at the same time.

You could do search for link matching Priority and Reliability and early return on a result, but if you don't find any result you need to search by Reliability only. And if you don't find any result either you pick any link.

The search uses a fold to keep track of all three kinds of matches.